docs/source/_notebooks/keras-text-classifiers.rst

```diff
@@ -0,0 +1,3685 @@
+
+Explaining Keras text classifier predictions with Grad-CAM
+==========================================================
+
+We can use ELI5 to explain text-based classifiers, i.e. models that take
+in a text and assign it to some class. Common examples include sentiment
+classification, labelling into categories, etc.
+
+The underlying method used is 'Grad-CAM'
+(https://arxiv.org/abs/1610.02391). This technique shows what parts of
+the input are the most important to the predicted result, by overlaying
+a "heatmap" over the original input.
+
+See also the tutorial for images
+(https://eli5.readthedocs.io/en/latest/tutorials/keras-image-classifiers.html).
+Certain sections such as 'removing softmax' and 'comparing different
+models' are relevant for text as well.
+
+Set up
+------
+
+First some imports
+
+.. code:: ipython3
+
+    import os
+    
+    import numpy as np
+    import pandas as pd
+    from IPython.display import display, HTML  # our explanations will be formatted in HTML
+    
+    # you may want to keep logging enabled when doing your own work
+    import logging
+    import tensorflow as tf
+    tf.get_logger().setLevel(logging.ERROR) # disable Tensorflow warnings for this tutorial
+    import warnings
+    warnings.simplefilter("ignore") # disable Keras warnings for this tutorial
+    import keras
+    from keras.preprocessing.sequence import pad_sequences
+    
+    import eli5
+
+
+.. parsed-literal::
+
+    Using TensorFlow backend.
+
+
+The rest of what we need in this tutorial is stored in the
+``tests/estimators`` package, whose source you can check for your own
+reference. You may need extra steps here to load your custom model and
+data.
+
+.. code:: ipython3
+
+    # we need to go back to top level in order to import some local ELI5 modules
+    
+    old = os.getcwd()
+    os.chdir('..')
+
+Explaining binary (sentiment) classifications
+---------------------------------------------
+
+In binary classification there is only one possible class to which a
+piece of text can either belong to or not. In sentiment classification,
+that class is whether the text is "positive" (belongs to the class) or
+"negative" (doesn't belong to the class).
+
+In this example we will have a recurrent model with word level
+tokenization, trained on the IMDB dataset
+(https://keras.io/datasets/#imdb-movie-reviews-sentiment-classification).
+The model has one output node that gives probabilities. Output close to
+1 is positive, and close to 0 is negative.
+
+See
+https://www.tensorflow.org/beta/tutorials/text/text\_classification\_rnn
+for a simple example of how to build such a model and prepare its input.
```

Comment:
If we have a notebook or script which trains this model, would be great to link to it here.